pkg/rc/manager.go

```diff
@@ -0,0 +1,183 @@
+	sessions <-chan string
+
+	// rcs managed by this manager
+	children map[fields.ID]managedRC
+
+	logger logging.Logger
+}
+
+type managedRC struct {
+	rc   ReplicationController
+	quit chan<- struct{}
+}
+
+func NewManager(
+	kpStore kpStore,
+	rcs rcstore.Store,
+	scheduler Scheduler,
+	labeler labels.Applicator,
+	sessions <-chan string,
+	logger logging.Logger,
+) *RCManager {
+	return &RCManager{
+		kpStore:   kpStore,
+		rcStore:   rcs,
+		scheduler: scheduler,
+		labeler:   labeler,
+		sessions:  sessions,
+		logger:    logger,
+		children:  make(map[fields.ID]managedRC),
+	}
+}
+
+// Start is a blocking function that monitors Consul for replication controllers.
+// The RCManager will attempt to claim replication controllers as they appear
+// and, if successful, will start goroutines for those replication controllers
+// to do their job. Closing the quit channel will cause this function to
+// return, but does not release locks.
+//
+// Start is not safe for concurrent execution. Do not execute multiple
+// concurrent instances of Start.
+func (rcm *RCManager) Start(quit <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	rcWatch, rcErr := rcm.rcStore.WatchNew(subQuit)
+	var session string
+
+	for {
+		select {
+		case <-quit:
+			// TODO: we have to release these locks as well to make sure others
+			// can capture them, otherwise the lock will persist until it expires
+			// on its own. in practice, we will also kill the consul session
+			// manager, so is this really a problem?
+			rcm.logger.WithField("session", session).Infoln("Halt requested, releasing replication controllers")
+			rcm.releaseChildren()
+			return
+		case session = <-rcm.sessions:
+			if session == "" {
+				// our session has expired, we must assume our locked children
+				// have all been released and that someone else may have
+				// claimed them by now
+				rcm.logger.NoFields().Errorln("Session expired, releasing replication controllers")
+				rcm.releaseChildren()
+			} else {
+				// a new session has been acquired - only happens after an
+				// expiration message, so len(children)==0. right now this is a
+				// noop, since the session has been set to its new value, but in
+				// the future we should use a refactored lock API and instantiate
+				// it here
+				rcm.logger.WithField("session", session).Infoln("Acquired new session")
+				// TODO: restart the watch so that you get updates right away?
+			}
+		case err := <-rcErr:
+			rcm.logger.WithError(err).Errorln("Could not read consul replication controllers")
+		case rcFields := <-rcWatch:
+			rcm.logger.WithField("n", len(rcFields)).Debugln("Received replication controller update")
+			if session == "" {
+				// we can't claim new nodes because our session is invalidated.
+				// raise an error and ignore this update
+				rcm.logger.NoFields().Infoln("Received replication controller update, but do not have session to acquire locks")
+				continue
```

Comment:
sure, this doesn't kill the loop. you'll just have to wait for the next batch of results come in, and then you'll start acquiring locks as usual again. The only problem is that it might take a while for the next batch if there are no changes to trigger the watch (you would have to wait for it to timeout). See my TODO comment about restarting the watch when a new session is obtained.